crates/red_knot_python_semantic/resources/mdtest/unpacking.md

```diff
@@ -536,6 +538,7 @@ for a, b in ((1, 2), ("a", "b")):
 # error: "Object of type `Literal[1]` is not iterable"
 # error: "Object of type `Literal[2]` is not iterable"
 # error: "Object of type `Literal[4]` is not iterable"
+# error: "Not enough values to unpack (expected 2, got 1)"
 for a, b in (1, 2, (3, "a"), 4, (5, "b"), "c"):
     reveal_type(a)  # revealed: Unknown | Literal[3, 5] | LiteralString
```

Comment:
I agree it would be better if this were just `Unknown | Literal[3, 5]`.

Reply:
I think to do this consistently would mean changing a bunch of other cases, though -- it would mean that e.g. for `(a, b, c) = (1, 2)` we would infer `Unknown` (or `Never` would also be reasonable) for all of `a`, `b`, `c`, rather than inferring `Literal[1]` for `a` and `Literal[2]` for `b`. That is, today we try to match up and infer as many types as we can, even if the overall unpacking is an error; this would mean changing that.

The advantage of the current approach is that it can catch more errors at once, assuming the parts that we are able to line up are actually lined up correctly. But that may not be a reasonable assumption; if you have a mis-alignment it could actually be early in the unpacking, in which case we are just inferring wrong types for many of the elements, leading to false positives in subsequent code.

It looks like our current behavior is more like pyright; mypy always just infers `Any` for all targets in a mis-aligned unpacking.

I do think that the mypy behavior is more principled and we should probably switch to that, though I also don't think it's high priority to make that change.

Reply:
Yes, I did that for the advantage that you've mentioned and I'm fine changing the behavior as it's an error case and should be less often in practice. I'll expand the original issue with this task.